File: project/drivers/driver/dart/lib/src/keepalive.dart
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import 'dart:async';

class KeepaliveConfig {
  int intervalMs;
  int maxIdleBeforeCheckMs;
  int validationTimeoutMs;

  KeepaliveConfig({
    this.intervalMs = 120000,
    this.maxIdleBeforeCheckMs = 600000,
    this.validationTimeoutMs = 5000,
  });
}

class KeepaliveTracker {
  final KeepaliveConfig config;
  int _lastActivity = DateTime.now().millisecondsSinceEpoch;

  KeepaliveTracker(this.config);

  void markActive() {
    _lastActivity = DateTime.now().millisecondsSinceEpoch;
  }

  bool needsValidation() {
    return DateTime.now().millisecondsSinceEpoch - _lastActivity > config.maxIdleBeforeCheckMs;
  }
}

typedef Pinger = Future<bool> Function();

class KeepaliveManager {
  final KeepaliveConfig config;
  final Map<String, KeepaliveTracker> _trackers = {};
  final Map<String, Pinger> _pingers = {};
  Timer? _timer;

  KeepaliveManager([KeepaliveConfig? config]) : config = config ?? KeepaliveConfig();

  void start() {
    if (_timer != null) return;
    _timer = Timer.periodic(Duration(milliseconds: config.intervalMs), (_) => _checkConnections());
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  KeepaliveTracker register(String connId, Pinger pinger) {
    final tracker = KeepaliveTracker(config);
    _trackers[connId] = tracker;
    _pingers[connId] = pinger;
    return tracker;
  }

  void unregister(String connId) {
    _trackers.remove(connId);
    _pingers.remove(connId);
  }

  Future<void> _checkConnections() async {
    for (final entry in _trackers.entries) {
      if (!entry.value.needsValidation()) continue;
      final pinger = _pingers[entry.key];
      if (pinger == null) continue;
      try {
        final ok = await _pingWithTimeout(pinger);
        if (ok) {
          entry.value.markActive();
        }
      } catch (_) {}
    }
  }

  Future<bool> _pingWithTimeout(Pinger pinger) async {
    return pinger().timeout(Duration(milliseconds: config.validationTimeoutMs), onTimeout: () => false);
  }
}

File: project/drivers/driver/node/src/keepalive.ts
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

export interface KeepaliveConfig {
  intervalMs?: number;
  maxIdleBeforeCheckMs?: number;
  validationTimeoutMs?: number;
}

export class KeepaliveTracker {
  private lastActivity = Date.now();

  constructor(private readonly config: Required<KeepaliveConfig>) {}

  markActive(): void {
    this.lastActivity = Date.now();
  }

  needsValidation(): boolean {
    return Date.now() - this.lastActivity > this.config.maxIdleBeforeCheckMs;
  }

  idleDurationMs(): number {
    return Date.now() - this.lastActivity;
  }
}

type Pinger = () => boolean | Promise<boolean>;

export class KeepaliveManager {
  private readonly config: Required<KeepaliveConfig>;
  private readonly trackers = new Map<string, KeepaliveTracker>();
  private readonly pingers = new Map<string, Pinger>();
  private timer?: NodeJS.Timeout;

  constructor(config: KeepaliveConfig = {}) {
    this.config = {
      intervalMs: config.intervalMs ?? 120_000,
      maxIdleBeforeCheckMs: config.maxIdleBeforeCheckMs ?? 600_000,
      validationTimeoutMs: config.validationTimeoutMs ?? 5_000,
    };
  }

  start(): void {
    if (this.timer) {
      return;
    }
    this.timer = setInterval(() => void this.checkConnections(), this.config.intervalMs);
    this.timer.unref?.();
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = undefined;
    }
  }

  register(connId: string, pinger: Pinger): KeepaliveTracker {
    const tracker = new KeepaliveTracker(this.config);
    this.trackers.set(connId, tracker);
    this.pingers.set(connId, pinger);
    return tracker;
  }

  unregister(connId: string): void {
    this.trackers.delete(connId);
    this.pingers.delete(connId);
  }

  private async checkConnections(): Promise<void> {
    for (const [connId, tracker] of this.trackers.entries()) {
      if (!tracker.needsValidation()) {
        continue;
      }
      const pinger = this.pingers.get(connId);
      if (!pinger) {
        continue;
      }
      try {
        const healthy = await this.pingWithTimeout(pinger, this.config.validationTimeoutMs);
        if (healthy) {
          tracker.markActive();
        }
      } catch {
        // ping failed; connection will be handled on next request
      }
    }
  }

  private async pingWithTimeout(pinger: Pinger, timeoutMs: number): Promise<boolean> {
    const result = Promise.resolve().then(() => pinger());
    const timeout = new Promise<boolean>((resolve) => {
      const handle = setTimeout(() => resolve(false), timeoutMs);
      handle.unref?.();
    });
    return Promise.race([result, timeout]);
  }
}
